src/util.tsx

```diff
@@ -1,42 +1,60 @@
+export type EffectConstructor = new (...args: any[]) => Effect
 
-export const resolveRef = (ref: Object3D | ObjectRef) => (isRef(ref) ? ref.current : ref)
+export type EffectProps<T extends EffectConstructor> = ReactThreeFiber.Node<
+  T extends Function ? T['prototype'] : InstanceType<T>,
+  T
+> &
+  ConstructorParameters<T>[0] & {
+    blendFunction?: BlendFunction
+    opacity?: number
+  }
 
-export const wrapEffect = <T extends new (...args: any[]) => Effect>(
-  effectImpl: T,
-  defaultBlendMode: BlendFunction = BlendFunction.NORMAL
-) =>
-  forwardRef<T, ConstructorParameters<typeof effectImpl>[0] & DefaultProps>(function Wrap(
-    { blendFunction, opacity, ...props }: React.PropsWithChildren<DefaultProps & ConstructorParameters<T>[0]>,
+let i = 0
+const components = new WeakMap<EffectConstructor, React.ExoticComponent<any>>()
+
+export const wrapEffect = <T extends EffectConstructor, P extends EffectProps<T>>(effect: T, defaults?: P) =>
+  /* @__PURE__ */ React.forwardRef<T, P>(function Effect(
+    { blendFunction = defaults?.blendFunction, opacity = defaults?.opacity, ...props },
     ref
   ) {
-    const invalidate = useThree((state) => state.invalidate)
-    const effect: Effect = useMemo(() => new effectImpl(props), [props])
-
-    useLayoutEffect(() => {
-      effect.blendMode.blendFunction = !blendFunction && blendFunction !== 0 ? defaultBlendMode : blendFunction
-      if (opacity !== undefined) effect.blendMode.opacity.value = opacity
-      invalidate()
-    }, [blendFunction, effect.blendMode, opacity])
-    return <primitive ref={ref} object={effect} dispose={null} />
+    let Component = components.get(effect)
+    if (!Component) {
+      const key = `@react-three/postprocessing/${effect.name}-${i++}`
+      extend({ [key]: effect })
+      components.set(effect, (Component = key as any))
+    }
+
+    const camera = useThree((state) => state.camera)
+    const args = React.useMemo(
+      () => [...((defaults?.args ?? []) as any[]), ...((props.args ?? [{ ...defaults, ...props }]) as any[])],
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      [JSON.stringify(props)]
+    )
+
+    return (
+      <Component
+        camera={camera}
+        blendMode-blendFunction={blendFunction}
+        blendMode-opacity={opacity}
+        {...props}
+        ref={ref}
+        args={args}
+      />
+    )
   })
 
-export const useVector2 = (props: any, key: string): Vector2 => {
+export const useVector2 = (props: any, key: string): THREE.Vector2 => {
   const vec: ReactThreeFiber.Vector2 = props[key]
-  return useMemo(() => {
-    if (vec instanceof Vector2) {
-      return new Vector2().set(vec.x, vec.y)
-    } else if (Array.isArray(vec)) {
-      const [x, y] = vec
-      return new Vector2().set(x, y)
-    }
+  return React.useMemo(() => {
+    if (typeof vec === 'number') return new THREE.Vector2(vec, vec)
+    else return new THREE.Vector2(...(vec as THREE.Vector2Tuple))
```

Comment:
I recommend also handling the case where `vec` is undefined and returning `undefined` respectively:

```js
if (!vec) return undefined
else if (vec ...)
```